applications/services/input/input.c

```diff
@@ -75,95 +63,135 @@ const char* input_get_type_name(InputType type) {
 
     for(size_t i = 0; i < input_pins_count; i++) {
-        furi_hal_gpio_add_int_callback(input_pins[i].gpio, input_isr, thread_id);
-        pin_states[i].pin = &input_pins[i];
-        pin_states[i].state = GPIO_Read(pin_states[i]);
-        pin_states[i].debounce = INPUT_DEBOUNCE_TICKS_HALF;
-        pin_states[i].press_timer =
-            furi_timer_alloc(input_press_timer_callback, FuriTimerTypePeriodic, &pin_states[i]);
-        pin_states[i].event_pubsub = event_pubsub;
-        pin_states[i].press_counter = 0;
+        if(instance->key_state[i].state != GPIO_Read(instance->key_state[i])) {
+            if(!instance->key_state[i].state) {
+                instance->key_state[i].repeat_count = 0;
+            }
+            instance->key_state[i].debounce_count = 1;
+            instance->key_state[i].sequence_counter = instance->sequence_counter++;
+            furi_event_loop_timer_start(instance->key_state[i].timer, INPUT_SRV_DEBOUNCE_TICKS);
+        }
     }
+    return true;
+}
 
-    while(1) {
-        bool is_changing = false;
-        for(size_t i = 0; i < input_pins_count; i++) {
-            bool state = GPIO_Read(pin_states[i]);
-            if(state) {
-                if(pin_states[i].debounce < INPUT_DEBOUNCE_TICKS) pin_states[i].debounce += 1;
-            } else {
-                if(pin_states[i].debounce > 0) pin_states[i].debounce -= 1;
-            }
+void input_isr_key(void* context) {
+    InputSrv* instance = context;
+    furi_semaphore_release(instance->input_semaphore);
+}
 
-            if(pin_states[i].debounce > 0 && pin_states[i].debounce < INPUT_DEBOUNCE_TICKS) {
-                is_changing = true;
-            } else if(pin_states[i].state != state) {
-                pin_states[i].state = state;
-
-                // Common state info
-                InputEvent event;
-                event.sequence_source = INPUT_SEQUENCE_SOURCE_HARDWARE;
-                event.key = pin_states[i].pin->key;
-
-                // Short / Long / Repeat timer routine
-                if(state) {
-                    pin_states[i].counter = counter++;
-                    event.sequence_counter = pin_states[i].counter;
-                    furi_timer_start(pin_states[i].press_timer, INPUT_PRESS_TICKS);
-                } else {
-                    event.sequence_counter = pin_states[i].counter;
-                    furi_timer_stop(pin_states[i].press_timer);
-                    while(furi_timer_is_running(pin_states[i].press_timer))
-                        furi_delay_tick(1);
-                    if(pin_states[i].press_counter < INPUT_LONG_PRESS_COUNTS) {
-                        event.type = InputTypeShort;
-                        furi_pubsub_publish(event_pubsub, &event);
-                    }
-                    pin_states[i].press_counter = 0;
-                }
+int32_t input_srv(void* p) {
+    UNUSED(p);
+    InputSrv* instance = malloc(sizeof(InputSrv));
```

Comment:
`InputSrv` is small, it should be perfectly fine to put it on the stack so as not to eat away from the user memory space.